File: backend/app/models/task.py
from sqlalchemy import Column, String, DateTime, Integer, Boolean, Float, Text
from sqlalchemy.sql import func
from app.db.session import Base

class Task(Base):
    __tablename__ = "tasks"

    id = Column(String, primary_key=True, index=True)
    userId = Column(String, name='user_id', nullable=False, index=True)
    date = Column(String, nullable=False, index=True)
    title = Column(String, nullable=False)
    status = Column(String, nullable=False)  # "todo", "in-progress", "done"
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updatedAt = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    deletedAt = Column(DateTime(timezone=True), nullable=True)

    # Time tracking fields
    spentTime = Column(Integer, default=0)  # Total time spent in minutes (single source of truth)
    timeLogs = Column(Text, nullable=True)  # JSON array of time log entries
    accumulated_time = Column(Integer, default=0)  # Legacy field for migration

    # Timer fields
    timer_running = Column(Boolean, default=False)
    timer_start = Column(DateTime(timezone=True), nullable=True)

    # Estimation and usefulness
    estimated_time = Column(Integer, nullable=True)  # Time in minutes
    is_useful = Column(Boolean, nullable=True)

File: domain/task.ts
export type TaskStatus = "todo" | "in-progress" | "done";

export interface TimeLogEntry {
  id: string;
  type: 'timer' | 'manual';
  minutes: number;
  createdAt: string;
}

export interface Task {
  id: string;
  userId: string;
  date: string;
  title: string;
  status: TaskStatus;
  createdAt: string;
  updatedAt?: string; // Optional for backward compatibility, auto-populated on save
  // Time tracking - spentTime is the single source of truth
  spentTime: number; // Total time spent in minutes
  timeLogs: TimeLogEntry[]; // Array of time log entries
  // Legacy field for migration (optional)
  accumulatedTime?: number;
  // Timer fields for time tracking
  timerRunning: boolean; // Whether timer is currently active
  timerStart: string | null; // ISO timestamp when timer was started, null if not running
  // Estimation and usefulness tracking
  estimatedTime: number | null; // Estimated time in minutes, null if not set
  isUseful: boolean | null; // true = useful, false = not useful, null = not categorized
  isCopiedToNextDay: boolean; // Star status for copying to next day
}
